nixos/tests/systemd-coredump-disabled.nix

```diff
@@ -0,0 +1,19 @@
+import ./make-test-python.nix ({ pkgs, ... }:
+
+{
+  name = "systemd-coredump-disabled";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ squalus ];
+  };
+
+  nodes.machine = { pkgs, lib, ... }: {
+    systemd.coredump.enable = false;
+    systemd.package = pkgs.systemd.override {
+      withCoredump = false;
+    };
+  };
+
+  testScript = ''
+    machine.wait_for_unit("multi-user.target")
```

Comment:
In the systemd-coredump disabled, I think we should also let the binary crash here and check for the default behaviour without the systemd sysctl snippet (core dump created in working directory).

You might want to add this logic into systemd-coredump, and override the config with the `specialisation` test feature.

Reply:
For the first point, that makes sense. I can add a test to check that the crashing program produces a coredump in the working directory, which is what we expect from the default sysctl config.

I don't understand the second point. Could you elaborate on this?

Reply:
If both tests run the binary, expect it to crash and then do some (slightly different) assertions, there's a lot of duplicate code. 

My proposal was to join the two files into a single `.nix` file, and use the `specialisation` test feature (grep `nixos/tests` for some other examples) to create two different VMs - one with coredump enabled, one with it being disabled (and `systemd.package` overridden to a version without support).

Then in the test script, boot both, assert `coredumpctl` showing the dump, and the other VM having produced the coredump in the working dir.

Reply:
Ok, that makes sense. I reworked it to use one test file and a specialisation for the disabled case. But I'm having trouble getting it to switch to a new configuration and then reboot. I think I need to run `nix-env` to set the system profile to the new specialisation, and then run `switch-to-configuration boot`, then restart the machine. Haven't been able to get this working yet.

Reply:
I think the sysctl's should also get applied on a simple switch, without a reboot.

Check https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/specialisation.nix for a simple example. It should be sufficient to do a `switch-to-configuration test` (which applies, but doesn't update bootloaders), without a reboot.

Reply:
The `switch-to-configuration test` operation never finishes. I have the intermediate work here: https://github.com/NixOS/nixpkgs/compare/master...squalus:nixpkgs:coredump-specialisation?expand=1

If I type the `switch-to-configuration test` command manually in the qemu vm, it works fine. But it hangs forever when invoked by the test script.

```
machine: must succeed: /run/current-system/specialisation/disabled/bin/switch-to-configuration test
machine # [    9.303393] nixos[912]: switching to system configuration /nix/store/y5hllpqa87mi4pjnx83lidwgm8kaj1pk-nixos-system-machine-22.11pre-git
machine # stopping the following units: audit.service, backdoor.service, logrotate-checkconf.service, mount-pstore.service, network-local-commands.service, network-setup.service, nscd.service, resolvconf.service, systemd-coredump.socket, systemd-modules-load.service, systemd-sysctl.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, systemd-udevd.service, systemd-update-done.service
machine # [    9.328545] systemd[1]: Stopping Kernel Auditing...
machine # [    9.337855] systemd[1]: Stopping backdoor.service...
machine # [    9.342305] systemd[1]: Stopped target Local File Systems.
machine # [    9.347503] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[929]: No rules
machine # [    9.354716] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: enabled 0
machine # [    9.357227] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: failure 1
machine # [    9.359783] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: pid 0
machine # [    9.362970] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: rate_limit 0
machine # [    9.364078] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: backlog_limit 64
machine # [    9.365135] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: lost 0
machine # [    9.366064] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: backlog 0
machine # [    9.367003] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: backlog_wait_time 60000
machine # [    9.368215] systemd[1]: audit.service: Deactivated successfully.
machine # [    9.369132] systemd[1]: Stopped Kernel Auditing.
machine # [    9.369992] systemd[1]: backdoor.service: Deactivated successfully.
machine # [    9.370990] systemd[1]: Stopped backdoor.service.
```

Reply:
Huh, that's odd. @grahamc, any idea on that?